File: internal/git/commit.go
package git

import (
	"fmt"
	"os/exec"
	"strings"
)

// Commit creates a git commit with the given message.
func Commit(dir string, message string) error {
	cmd := exec.Command("git", "-C", dir, "commit", "-m", message)
	out, err := cmd.CombinedOutput()
	if err != nil {
		return fmt.Errorf("git commit: %s: %w", strings.TrimSpace(string(out)), err)
	}
	return nil
}

// Push pushes the current branch to origin.
func Push(dir string) error {
	cmd := exec.Command("git", "-C", dir, "push", "-u", "origin", "HEAD")
	out, err := cmd.CombinedOutput()
	if err != nil {
		return fmt.Errorf("git push: %s: %w", strings.TrimSpace(string(out)), err)
	}
	return nil
}

// CreatePR opens a pull request using the gh CLI.
func CreatePR(dir string, title string, body string) error {
	args := []string{"pr", "create", "--title", title, "--body", body}
	cmd := exec.Command("gh", args...)
	cmd.Dir = dir
	out, err := cmd.CombinedOutput()
	if err != nil {
		return fmt.Errorf("gh pr create: %s: %w", strings.TrimSpace(string(out)), err)
	}
	return nil
}

// StagedDiff returns the staged diff (--cached).
func StagedDiff(dir string) (string, error) {
	cmd := exec.Command("git", "-C", dir, "diff", "--cached")
	out, err := cmd.Output()
	if err != nil {
		return "", fmt.Errorf("git diff --cached: %w", err)
	}
	return string(out), nil
}

// HasStagedChanges returns true if there are staged changes.
func HasStagedChanges(dir string) bool {
	cmd := exec.Command("git", "-C", dir, "diff", "--cached", "--quiet")
	return cmd.Run() != nil // exit 1 means there are changes
}

// HasUnpushedCommits returns true if the current branch has commits not yet pushed.
func HasUnpushedCommits(dir string) bool {
	cmd := exec.Command("git", "-C", dir, "log", "--oneline", "@{u}..HEAD")
	out, err := cmd.Output()
	if err != nil {
		// No upstream set — treat as having unpushed commits.
		return true
	}
	return len(strings.TrimSpace(string(out))) > 0
}

// HasOpenPR returns true if the current branch already has an open PR.
func HasOpenPR(dir string) bool {
	cmd := exec.Command("gh", "pr", "view", "--json", "state", "-q", ".state")
	cmd.Dir = dir
	out, err := cmd.Output()
	if err != nil {
		return false
	}
	return strings.TrimSpace(string(out)) == "OPEN"
}

// BranchDiff returns the diff of the current branch vs the default branch.
func BranchDiff(dir string) (string, error) {
	// Find the merge base with the default branch.
	base := exec.Command("gh", "repo", "view", "--json", "defaultBranchRef", "-q", ".defaultBranchRef.name")
	base.Dir = dir
	baseOut, err := base.Output()
	defaultBranch := strings.TrimSpace(string(baseOut))
	if err != nil || defaultBranch == "" {
		defaultBranch = "main"
	}

	cmd := exec.Command("git", "-C", dir, "diff", defaultBranch+"...HEAD")
	out, err := cmd.Output()
	if err != nil {
		return "", fmt.Errorf("git diff branch: %w", err)
	}
	return string(out), nil
}

// BranchLog returns the commit log of the current branch vs the default branch.
func BranchLog(dir string) (string, error) {
	base := exec.Command("gh", "repo", "view", "--json", "defaultBranchRef", "-q", ".defaultBranchRef.name")
	base.Dir = dir
	baseOut, err := base.Output()
	defaultBranch := strings.TrimSpace(string(baseOut))
	if err != nil || defaultBranch == "" {
		defaultBranch = "main"
	}

	cmd := exec.Command("git", "-C", dir, "log", "--oneline", defaultBranch+"..HEAD")
	out, err := cmd.Output()
	if err != nil {
		return "", fmt.Errorf("git log branch: %w", err)
	}
	return string(out), nil
}

File: src/git/commit.rs
use anyhow::{Result, bail};
use tokio::process::Command;

pub async fn commit(dir: &str, message: &str) -> Result<()> {
    let output = Command::new("git")
        .args(["-C", dir, "commit", "-m", message])
        .output()
        .await?;
    if !output.status.success() {
        bail!(
            "git commit: {}",
            String::from_utf8_lossy(&output.stderr)
        );
    }
    Ok(())
}

pub async fn push(dir: &str) -> Result<()> {
    let output = Command::new("git")
        .args(["-C", dir, "push", "-u", "origin", "HEAD"])
        .output()
        .await?;
    if !output.status.success() {
        bail!(
            "git push: {}",
            String::from_utf8_lossy(&output.stderr)
        );
    }
    Ok(())
}

pub async fn create_pr(dir: &str, title: &str, body: &str) -> Result<()> {
    let output = Command::new("gh")
        .args(["pr", "create", "--title", title, "--body", body])
        .current_dir(dir)
        .output()
        .await?;
    if !output.status.success() {
        bail!(
            "gh pr create: {}",
            String::from_utf8_lossy(&output.stderr)
        );
    }
    Ok(())
}

pub async fn has_staged_changes(dir: &str) -> bool {
    Command::new("git")
        .args(["-C", dir, "diff", "--cached", "--quiet"])
        .output()
        .await
        .map(|o| !o.status.success()) // exit 1 = changes exist
        .unwrap_or(false)
}

pub async fn has_unstaged_changes(dir: &str) -> bool {
    Command::new("git")
        .args(["-C", dir, "status", "--porcelain"])
        .output()
        .await
        .map(|o| !String::from_utf8_lossy(&o.stdout).trim().is_empty())
        .unwrap_or(false)
}

pub async fn has_unpushed_commits(dir: &str) -> bool {
    Command::new("git")
        .args(["-C", dir, "log", "--oneline", "@{u}..HEAD"])
        .output()
        .await
        .map(|o| {
            if !o.status.success() {
                true // no upstream = treat as unpushed
            } else {
                !String::from_utf8_lossy(&o.stdout).trim().is_empty()
            }
        })
        .unwrap_or(true)
}

pub async fn has_open_pr(dir: &str) -> bool {
    Command::new("gh")
        .args(["pr", "view", "--json", "state", "-q", ".state"])
        .current_dir(dir)
        .output()
        .await
        .map(|o| String::from_utf8_lossy(&o.stdout).trim() == "OPEN")
        .unwrap_or(false)
}

pub async fn staged_diff(dir: &str) -> Result<String> {
    let output = Command::new("git")
        .args(["-C", dir, "diff", "--cached"])
        .output()
        .await?;
    Ok(String::from_utf8_lossy(&output.stdout).to_string())
}

pub async fn branch_diff(dir: &str) -> Result<String> {
    let default_branch = get_default_branch_via_gh(dir).await;
    let output = Command::new("git")
        .args(["-C", dir, "diff", &format!("{default_branch}...HEAD")])
        .output()
        .await?;
    if !output.status.success() {
        bail!(
            "git diff branch: {}",
            String::from_utf8_lossy(&output.stderr)
        );
    }
    Ok(String::from_utf8_lossy(&output.stdout).to_string())
}

pub async fn branch_log(dir: &str) -> Result<String> {
    let default_branch = get_default_branch_via_gh(dir).await;
    let output = Command::new("git")
        .args(["-C", dir, "log", "--oneline", &format!("{default_branch}..HEAD")])
        .output()
        .await?;
    if !output.status.success() {
        bail!(
            "git log branch: {}",
            String::from_utf8_lossy(&output.stderr)
        );
    }
    Ok(String::from_utf8_lossy(&output.stdout).to_string())
}

async fn get_default_branch_via_gh(dir: &str) -> String {
    Command::new("gh")
        .args([
            "repo",
            "view",
            "--json",
            "defaultBranchRef",
            "-q",
            ".defaultBranchRef.name",
        ])
        .current_dir(dir)
        .output()
        .await
        .map(|o| {
            let s = String::from_utf8_lossy(&o.stdout).trim().to_string();
            if s.is_empty() { "main".to_string() } else { s }
        })
        .unwrap_or_else(|_| "main".to_string())
}
